### PR TITLE
Issue #2656818 by vasike, smccabe, bojanz: Implement optimistic locking for orders

### PR DESCRIPTION
**Pull request by vasike**

### CODE REVIEW

File: modules/cart/commerce_cart.module [hunk 1]
```diff
@@ -108,6 +109,55 @@ function commerce_cart_commerce_order_delete(OrderInterface $order) {
...
+  }
+}
+
+/**
+ * Validate handler for Views commerce cart forms.
+ */
+function commerce_cart_commerce_cart_form_validate($form, FormStateInterface $form_state) {
+  // Re-load the order.
+  /** @var \Drupal\commerce_order\Entity\OrderInterface $order */
+  $order = Order::load($form_state->get('order_id'));
```
Comment (mglaman) [inline modules/cart/commerce_cart.module]: Should we use `loadUnchanged` from storage to ensure it's fresh?

File: modules/cart/commerce_cart.module [hunk 2]
```diff
@@ -138,6 +188,10 @@ function commerce_cart_form_commerce_order_form_alter(array &$form, FormStateInt
       // Move the cart element to the bottom of the meta sidebar container.
       $form['cart']['#group'] = 'meta';
       $form['cart']['#weight'] = 100;
+      // Warning it the order is cart and it was modified in the last hour.
+      if (!empty($order->get('cart')->value) && (REQUEST_TIME - $order->getChangedTime()) < 3600) {
```
Comment (mglaman) [inline modules/cart/commerce_cart.module]: Time service

File: modules/order/src/Exception/ExceptionInterface.php [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+namespace Drupal\commerce_order\Exception;
+
+/**
+ * Exception interface for all exceptions thrown by the Order module.
+ */
+interface ExceptionInterface {}
```
Comment (mglaman) [inline modules/order/src/Exception/ExceptionInterface.php]: OrderExceptionInterface?

File: modules/order/src/Form/OrderForm.php [hunk 4]
```diff
@@ -61,6 +61,11 @@ public function form(array $form, FormStateInterface $form_state) {
       '#type' => 'hidden',
       '#default_value' => $order->getChangedTime(),
     ];
+    // Version must be sent to the client, for later overwrite error checking.
+    $form['version'] = [
+      '#type' => 'hidden',
```
Comment (mglaman) [inline modules/order/src/Form/OrderForm.php:74]: Can't we use #value?

File: modules/cart/commerce_cart.module (hunk 1 above)
Comment (vasike) [reply modules/cart/commerce_cart.module]: indeed. i'll do

File: modules/cart/commerce_cart.module (hunk 2 above)
Comment (vasike) [reply modules/cart/commerce_cart.module]: i did intentionally as i think soon we'll switch to 8.3 and use core time.
however i'll switch to Commerce Time till then. thanks

File: modules/order/src/Exception/ExceptionInterface.php (hunk 3 above)
Comment (vasike) [reply modules/order/src/Exception/ExceptionInterface.php]: agree

File: modules/order/src/Form/OrderForm.php (hunk 4 above)
Comment (vasike) [reply modules/order/src/Form/OrderForm.php:74]: it seems it won't work with value. we need to send it to client.

**Comment by bojanz** [comment]
An updated patch is being posted to the issue.

**Pull request closed without merging**